src/NHibernate/Driver/OracleManagedDataClientDriver.cs

```diff
@@ -47,8 +48,31 @@ public OracleManagedDataClientDriver()
 			oracleDbTypeRefCursor = Enum.Parse(oracleDbTypeEnum, "RefCursor");
 			oracleDbTypeXmlType = Enum.Parse(oracleDbTypeEnum, "XmlType");
 			oracleDbTypeBlob = Enum.Parse(oracleDbTypeEnum, "Blob");
+			_oracleDbTypeNVarchar2 = Enum.Parse(oracleDbTypeEnum, "NVarchar2");
 		}
 
+		/// <summary>
+		/// By default, ODP.Net maps <see cref="DbType.String"/> to <c>OracleDbType.Varchar2</c>. If you need it to be mapped
+		/// to <c>OracleDbType.NVarchar2</c> instead, set this property to <see langword="true"/>.
+		/// </summary>
+		/// <remarks>
+		/// <para>
+		/// This is not a bug. Varchar2 with Oracle can be Unicode, depending on the server configuration. If the server is
+		/// configured for having it Unicode, Varchar2 should be used for Unicode string. NVarchar2 is there for allowing
+		/// storing Unicode string on server not configured for Unicode.
+		/// http://docs.oracle.com/cd/E51173_01/win.122/e17732/featOraCommand.htm#i1007432
+		/// </para>
+		/// <para>
+		/// So the default NHibernate choice in Oracle dialects to map <see cref="DbType.String"/> to <c>nvarchar2</c> is a bit
+		/// unfortunate, because this is not the default mapping for Oracle. For tests, since the database is created by NHibernate,
+		/// string columns get typed <c>nvarchar2</c>, but parameters are still typed <c>varchar2</c>, causing an
+		/// <c>ORA-12704: character set mismatch</c> for some queries. This being not even reported, I guess most NHibernate Oracle
+		/// users handles their tables themselves and types them with <c>varchar2</c>, avoiding the trouble. So changing the behavior
+		/// for parameters type would be a major breaking change.
+		/// </para>
+		/// </remarks>
+		public bool UseNVarchar2ForStringParameter { get; set; }
```

Comment:
I am convinced that NHibernate is wrong assuming Oracle Varchar2 is meant for single byte character encoding. See this [Oracle](https://docs.oracle.com/cd/B19306_01/server.102/b14225/ch6unicode.htm#CACHCAHF) page.

They have not implemented Unicode as a new string type, but simply as a new encoding, usable with any already existing string and character types.

The default recommended way of using Unicode with Oracle is to use a varchar2 with an adequate database encoding.

> You can create a Unicode database that enables you to store UTF-8 encoded characters as SQL `CHAR` datatypes (`CHAR`, `VARCHAR2`, `CLOB`, and `LONG`).

Using the `N` prefixed types are second option, notably meant for people wanting to transition incrementally to Unicode instead of globally switching to it.

> If you prefer to implement Unicode support incrementally or if you need to support multilingual data only in certain columns, then you can store Unicode data in either the UTF-16 or UTF-8 encoding form in SQL `NCHAR` datatypes (`NCHAR`, `NVARCHAR2`, and `NCLOB`). The SQL `NCHAR` datatypes are called Unicode datatypes because they are used only for storing Unicode data.

This effectively means Oracle has two Unicode support models. And NHibernate implements the one which looks like being the less used. If we really want to fix this mess, we need to give an option for letting the user tell which model he uses. This could be that `UseNVarchar2ForStringParameter` option, but for completeness, it needs to be taken into account by the dialect too.

The first recommended one, by the way, does not support anymore single byte encoding at all.